### PR TITLE
Allow ARM Linux builds

### PR DESCRIPTION
**Pull request by ethankhall**
This change-set should allow ARM Linux builds for Sorbet artifacts.

### Motivation

https://github.com/sorbet/sorbet/issues/4119

### Test plan

All existing tests should work against the repo. Changes to buildkite are needed to still, and require work from Sorbet team to enable.

See included automated tests.

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -17,14 +17,13 @@ load("@com_grail_bazel_toolchain//toolchain:rules.bzl", "llvm_toolchain")
 llvm_toolchain(
     name = "llvm_toolchain_12_0_0",
     absolute_paths = True,
-    llvm_mirror_prefixes = [
-        "https://sorbet-deps.s3-us-west-2.amazonaws.com/",
-        "https://artifactory-content.stripe.build/artifactory/github-archives/llvm/llvm-project/releases/download/llvmorg-",
-        "https://github.com/llvm/llvm-project/releases/download/llvmorg-",
-    ],
```
Comment (jez) [inline WORKSPACE:24]: Why is this being deleted?

This is required to get Sorbet to build on Stripe's internal network, and cannot be deleted.

Comment (ethankhall) [reply WORKSPACE:24]: I had issues with it when prototyping, I forgot to revert it. Will add back in.

File: third_party/externals.bzl [hunk 2]
```diff
@@ -168,12 +168,15 @@ def register_sorbet_dependencies():
...
+
     http_archive(
         name = "com_grail_bazel_toolchain",
-        urls = _github_public_urls("sorbet/bazel-toolchain/archive/a685e1e6bd1e7cc9a5b84f832539585bb68d8ab4.zip"),
-        sha256 = "90c59f14cada755706a38bdd0f5ad8f0402cbf766387929cfbee9c3f1b4c82d7",
-        strip_prefix = "bazel-toolchain-a685e1e6bd1e7cc9a5b84f832539585bb68d8ab4",
+        sha256 = LLVM_BAZEL_TOOLCHAIN_SHA,
+        strip_prefix = "bazel-toolchain-{tag}".format(tag = LLVM_BAZEL_TOOLCHAIN_TAG),
+        canonical_id = LLVM_BAZEL_TOOLCHAIN_TAG,
+        url = "https://github.com/grailbio/bazel-toolchain/archive/{tag}.tar.gz".format(tag = LLVM_BAZEL_TOOLCHAIN_TAG),
```
Comment (jez) [inline third_party/externals.bzl:179]: There are custom changes in the `sorbet/bazel-toolchain` fork of `grailbio/bazel-toolchain` that we can't simply throw away. We have to backport the changes we need from upstream, or make sure that the changes that have been made in Sorbet's fork are applied upstream.

File: third_party/externals.bzl [hunk 3]
```diff
@@ -196,15 +199,6 @@ def register_sorbet_dependencies():
...
     )
 
-    # optimized version of blake2 hashing algorithm
-    http_archive(
-        name = "com_github_blake2_libb2",
-        urls = _github_public_urls("BLAKE2/libb2/archive/fa83ddbe179912e9a7a57edf0333b33f6ff83056.zip"),
-        sha256 = "dd25f7ac53371c2a15761fc1689d04de2ff948ac7e213a10d13961e24b0c9ae6",
-        build_file = "@com_stripe_ruby_typer//third_party:libb2.BUILD",
-        strip_prefix = "libb2-fa83ddbe179912e9a7a57edf0333b33f6ff83056",
-    )
```
Comment (jez) [inline third_party/externals.bzl:206]: This library is a faster version of the blake2 algorithm on platforms that support it. Why has it been deleted?

**Review by jez — CHANGES_REQUESTED** [review]
^

File: bazel [hunk 4]
```diff
@@ -55,6 +55,7 @@ processor_name="$(uname -m)"
 bazel_installer_platform="${kernel_name}-${processor_name}"
 case "$bazel_installer_platform" in
   linux-x86_64) ;;
+  linux-arm64) ;;
```
Comment (jez) [inline bazel:58]: Why do we not need to add a checksum to the bazel installer at the end of the `WORKSPACE` file like we do for all the other installers?

https://github.com/sorbet/sorbet/blob/52da7a7809556280402c918409cf2de486db9d5c/WORKSPACE#L62-L68

File: .bazelrc [hunk 5]
```diff
@@ -12,7 +12,7 @@ common:test --test_env=GTEST_INSTALL_FAILURE_SIGNAL_HANDLER=1
 ##
 ##  Custom toolchain.
 ##
-build --crosstool_top=@llvm_toolchain_12_0_0//:toolchain --copt=-D_LIBCPP_ENABLE_NODISCARD
```
Comment (jez) [inline .bazelrc:15]: Why do we not need this `D_LIBCPP_ENABLE_NODISCARD` definition anymore?

Comment (ethankhall) [reply .bazelrc:15]: No idea. I was following suggestions around crosstool_top. I can add it back in.

File: third_party/externals.bzl (hunk 2 above)
Comment (ethankhall) [reply third_party/externals.bzl:179]: Would you recommend rebasing your brach on the upstream and tryin it?

The upstream brach was able to build arm linux artifacts and we were able to use the sucessfully.

**Comment by ksapozhnikov** [comment]
Any progress on this?

**Comment by OAuthBringer** [comment]
🙏

**Comment by kaibadash** [comment]
🙏💪

**Comment by jeffsaremi** [comment]
Could we get this moving forward please? thanks

**Comment by jeffsaremi** [comment]
@ ethankhall How did you get this to build on arm64? I encounter problems every step of the way from bazel install to getting sorbet build using the right toolchain

**Comment by kivikakk** [comment]
Hi friends! I've recently had a go at this, building on @ethankhall's work:

https://github.com/sorbet/sorbet/compare/master...radiopaedia:sorbet:ethankhall/arm-build

I've rebased this branch and cleaned up some issues. Importantly, I've also rebased sorbet/bazel-toolchain on its upstream, which is used in this branch:

https://github.com/grailbio/bazel-toolchain/compare/master...radiopaedia:bazel-toolchain:sorbet

This was very much "best effort". I don't know Bazel at all, and while I've done my best to rework the changes made, it's probably broken some things I'm not aware of.

That said, it builds on linux-aarch64. I used [Bazelisk](https://github.com/bazelbuild/bazelisk) to get a working Bazel of the right version. Create the following `.bazeliskrc`:

```
USE_BAZEL_VERSION=5.2.0
BAZELISK_SKIP_WRAPPER=1
```

You will need to install deps; try `apt install git build-essential curl perl unzip python autoconf libffi-dev libgmp-dev libtinfo5 libtinfo-dev python python3 git libssl-dev libreadline-dev zlib1g-dev` or similar.

Then run:

```console
$ time bazelisk build //main:sorbet --config=release-linux-aarch64 --verbose_failures
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
WARNING: The following configs were expanded more than once: [static-libs]. For repeatable flags, repeats are counted twice and may lead to unexpected behavior.
WARNING: Download from https://sorbet-deps.s3-us-west-2.amazonaws.com/12.0.0/clang+llvm-12.0.0-aarch64-linux-gnu.tar.xz failed: class java.io.IOException GET returned 403 Forbidden
WARNING: Download from https://artifactory-content.stripe.build/artifactory/github-archives/llvm/llvm-project/releases/download/llvmorg-12.0.0/clang+llvm-12.0.0-aarch64-linux-gnu.tar.xz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Unknown host: artifactory-content.stripe.build
WARNING: Download from https://mirror.bazel.build/www.colm.net/files/ragel/ragel-6.10.tar.gz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://mirror.bazel.build/ftp.gnu.org/gnu/m4/m4-1.4.18.tar.xz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://mirror.bazel.build/github.com/jmillikin/rules_m4/releases/download/v0.1/m4-gnulib-788db09a9f88abbef73c97e8d7291c40455336d8.tar.xz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://mirror.bazel.build/ftp.gnu.org/gnu/bison/bison-3.3.2.tar.xz failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: Download from https://mirror.bazel.build/github.com/jmillikin/rules_bison/releases/download/v0.1/bison-gnulib-788db09a9f88abbef73c97e8d7291c40455336d8.tar.xz failed: class java.io.FileNotFoundException GET returned 404 Not Found
INFO: Analyzed target //main:sorbet (150 packages loaded, 3733 targets configured).
INFO: Found 1 target...
Target //main:sorbet up-to-date:
  bazel-bin/main/sorbet
INFO: Elapsed time: 852.603s, Critical Path: 151.18s
INFO: 1401 processes: 227 internal, 1174 local.
INFO: Build completed successfully, 1401 total actions

________________________________________________________
Executed in  852.69 secs   fish           external 
   usr time  1387.47 millis    0.00 micros  1387.47 millis 
   sys time  472.97 millis  1352.00 micros  471.62 millis 

$ bazel-bin/main/sorbet -e '42 + "hello"'
-e:1: Expected Integer but found String("hello") for argument arg0 https://srb.help/7002
     1 |42 + "hello"
             ^^^^^^^
  Expected Integer for argument arg0 of method Integer#+:
    https://github.com/sorbet/sorbet/tree/7c04064e6cfed94a962516f9c12b34d497875e4c/rbi/core/integer.rbi#L148:
     148 |        arg0: Integer,
                  ^^^^
  Got String("hello") originating from:
    -e:1:
     1 |42 + "hello"
             ^^^^^^^
Errors: 1
$
```

I'm moving onto building the gems now, having gotten a binary reliably produced.

**Comment by jeffsaremi** [comment]
Thanks @kivikakk
What is the significance of the bazel-toolchaing fixes? and at what point do they come into play? Otherwise , without that my build fails with
```
ERROR: /root/.cache/bazel/_bazel_root/8c069df52082beee3c95ca17836fb8e2/external/jemalloc/BUILD.bazel:65:8: Executing genrule @jemalloc//:jemalloc_genrule failed: (Exit 1): bash failed: error executing command /bin/bash -c ... (remaining 1 argument skipped)
autoconf
./configure --enable-autogen --without-export --disable-shared --enable-static
checking for xsltproc... false
checking for gcc... sh /root/.cache/bazel/_bazel_root/8c069df52082beee3c95ca17836fb8e2/execroot/com_stripe_ruby_typer/external/llvm_toolchain_12_0_0/bin/cc_wrapper.sh
checking whether the C compiler works... no
configure: error: in `/root/.cache/bazel/_bazel_root/8c069df52082beee3c95ca17836fb8e2/execroot/com_stripe_ruby_typer/external/jemalloc':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

**Comment by kivikakk** [comment]
Hi @jeffsaremi! What's the content of `/root/.cache/bazel/_bazel_root/8c069df52082beee3c95ca17836fb8e2/execroot/com_stripe_ruby_typer/external/jemalloc/config.log` at this stage?

**Comment by jeffsaremi** [comment]
What is this PR waiting for? @jez

**Comment by kivikakk** [comment]
Fwiw, my branch still doesn't build cleanly on all arches, so I think there's still a fair bit of work yet.

**Comment by jeffsaremi** [comment]
@kivikakk Thanks very much!

**Comment by jeffsaremi** [comment]
@ethankhall @kivikakk Did you have to deal with this issue ever?
```
ERROR: /root/.cache/bazel/_bazel_root/7c8a07e8b28b5cb27445d0d91832b184/external/m4_v1.4.18/gnulib/BUILD.bazel:192:11: Compiling gnulib/lib/c-stack.c [for host] failed: (Exit 1): clang failed: error executing command
  (cd /root/.cache/bazel/_bazel_root/7c8a07e8b28b5cb27445d0d91832b184/execroot/com_stripe_ruby_typer && \
  exec env - \
    PATH=/bin:/usr/bin:/usr/local/bin \
    PWD=/proc/self/cwd \
  /root/.cache/bazel/_bazel_root/7c8a07e8b28b5cb27445d0d91832b184/external/llvm_toolchain_12_0_0/bin/clang -U_FORTIFY_SOURCE -fstack-protector -fno-omit-frame-pointer -fcolor-diagnostics -Wall -Wthread-safety -Wself-assign -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections -MD -MF bazel-out/host/bin/external/m4_v1.4.18/gnulib/_objs/gnulib/c-stack.d -iquote external/m4_v1.4.18 -iquote bazel-out/host/bin/external/m4_v1.4.18 -Ibazel-out/host/bin/external/m4_v1.4.18/gnulib/_virtual_includes/gnulib -isystem external/m4_v1.4.18/gnulib/config-linux -isystem bazel-out/host/bin/external/m4_v1.4.18/gnulib/config-linux -no-canonical-prefixes -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' '-fdebug-prefix-map=/root/.cache/bazel/_bazel_root/7c8a07e8b28b5cb27445d0d91832b184/external/llvm_toolchain_12_0_0/=external/llvm_toolchain_12_0_0/' -g0 -O1 -DFORCE_DEBUG -DHAVE_CONFIG_H -c external/m4_v1.4.18/gnulib/lib/c-stack.c -o bazel-out/host/bin/external/m4_v1.4.18/gnulib/_objs/gnulib/c-stack.o)
# Configuration: f1ffb6bf31ea1f4000741c9702f9ed066354e3192b10bc95b9e17b6f03ee9d4e
# Execution platform: @local_config_platform//:host
external/m4_v1.4.18/gnulib/lib/c-stack.c:55:26: error: function-like macro 'sysconf' is not defined
#elif HAVE_LIBSIGSEGV && SIGSTKSZ < 16384
                         ^
/usr/include/aarch64-linux-gnu/bits/sigstksz.h:28:19: note: expanded from macro 'SIGSTKSZ'
# define SIGSTKSZ sysconf (_SC_SIGSTKSZ)
                  ^
external/m4_v1.4.18/gnulib/lib/c-stack.c:139:8: error: fields must have a constant size: 'variable length array in structure' extension will never be supported
  char buffer[SIGSTKSZ];
       ^
  ```

**Comment by jeffsaremi** [comment]
@kivikakk I cloned and ran your version in a ubuntu 22 container with all the dependecies you mentioned. The only exception was that I compiled bazel 5.2.0 on the box instead of using bazelisk.
I got this:
```
# bazel build //main:sorbet --config=release-linux-aarch64 --verbose_failures
...
ERROR: /root/.cache/bazel/_bazel_root/7c8a07e8b28b5cb27445d0d91832b184/external/jemalloc/BUILD.bazel:65:8: Executing genrule @jemalloc//:jemalloc_genrule failed: (Exit 1): bash failed: error executing command
  (cd /root/.cache/bazel/_bazel_root/7c8a07e8b28b5cb27445d0d91832b184/execroot/com_stripe_ruby_typer && \
  exec env - \
    PATH=/bin:/usr/bin:/usr/local/bin \
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh;
  absolutize() {
  ...
  Target //main:sorbet failed to build
INFO: Elapsed time: 1.096s, Critical Path: 0.86s
INFO: 18 processes: 17 internal, 1 local.
FAILED: Build did NOT complete successfull
  ```

**Comment by kivikakk** [comment]
Hi @jeffsaremi — I don't have the time to support you here.

**Comment by tonybruess** [comment]
@kivikakk thanks for your hard work. Would you be willing to open a PR from your branch since this PR seems semi-abandoned?

**Comment by kivikakk** [comment]
@tonybruess Of course! On it.

**Pull request closed without merging**